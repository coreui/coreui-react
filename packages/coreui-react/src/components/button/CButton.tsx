import React, { ElementType, forwardRef } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import { CLink, CLinkProps } from '../link/CLink'

import { PolymorphicRefForwardingComponent } from '../../helpers'
import { colorPropType } from '../../props'
import type { Colors, Shapes } from '../../types'

export interface CButtonProps extends Omit<CLinkProps, 'size'> {
  /**
   * Toggle the active state for the component.
   */
  active?: boolean
  /**
   * Component used for the root node. Either a string to use a HTML element or a component.
   */
  as?: ElementType
  /**
   * A string of all className you want applied to the base component.
   */
  className?: string
  /**
   * Sets the color context of the component to one of CoreUIâ€™s themed colors.
   *
   * @type 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'dark' | 'light' | string
   */
  color?: Colors
  /**
   * Toggle the disabled state for the component.
   */
  disabled?: boolean
  /**
   * The href attribute specifies the URL of the page the link goes to.
   */
  href?: string
  /**
   * The role attribute describes the role of an element in programs that can make use of it, such as screen readers or magnifiers.
   */
  role?: string
  /**
   * Select the shape of the component.
   *
   * @type 'rounded' | 'rounded-top' | 'rounded-end' | 'rounded-bottom' | 'rounded-start' | 'rounded-circle' | 'rounded-pill' | 'rounded-0' | 'rounded-1' | 'rounded-2' | 'rounded-3' | string
   */
  shape?: Shapes
  /**
   * Size the component small or large.
   */
  size?: 'sm' | 'lg'
  /**
   * Specifies the type of button. Always specify the type attribute for the `<button>` element.
   * Different browsers may use different default types for the `<button>` element.
   */
  type?: 'button' | 'submit' | 'reset'
  /**
   * Set the button variant to an outlined button or a ghost button.
   */
  variant?: 'outline' | 'ghost'
}

export const CButton: PolymorphicRefForwardingComponent<'button', CButtonProps> = forwardRef<
  HTMLButtonElement | HTMLAnchorElement,
  CButtonProps
>(
  (
    { children, as = 'button', className, color, shape, size, type = 'button', variant, ...rest },
    ref
  ) => {
    return (
      <CLink
        as={rest.href ? 'a' : as}
        {...(!rest.href && { type: type })}
        className={classNames(
          'btn',
          variant && color ? `btn-${variant}-${color}` : `btn-${variant}`,
          {
            [`btn-${color}`]: color && !variant,
            [`btn-${size}`]: size,
          },
          shape,
          className
        )}
        {...rest}
        ref={ref}
      >
        {children}
      </CLink>
    )
  }
)

CButton.propTypes = {
  as: PropTypes.elementType,
  children: PropTypes.node,
  className: PropTypes.string,
  color: colorPropType,
  shape: PropTypes.string,
  size: PropTypes.oneOf(['sm', 'lg']),
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  variant: PropTypes.oneOf(['outline', 'ghost']),
}

CButton.displayName = 'CButton'
