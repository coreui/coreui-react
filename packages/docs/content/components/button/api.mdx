---
title: React Button Component API
name: Button API
description: Explore the API reference for the React Button component and discover how to effectively utilize its props for customization.
route: /components/button/
---

import CButtonAPI from '../../api/CButton.api.mdx'

## CButton

<CButtonAPI />
