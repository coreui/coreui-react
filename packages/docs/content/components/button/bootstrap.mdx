---
title: React Bootstrap Button Component
name: Buttons with Bootstrap Styling
description: Learn how to use CoreUI’s React Button component with Bootstrap styling to build interactive, accessible, and responsive buttons for forms, tables, modals, and more.
route: /components/button/
bootstrap_component: true
---

## How to use React Bootstrap Button component

CoreUI provides a set of prebuilt React Button components styled with Bootstrap. Buttons signal actions to users and can initiate events both in the UI and behind the scenes.

<ExampleSnippet className="theme-bootstrap" component="ButtonExample" componentName="React Bootstrap Button" />

<Callout color="warning" type="colorAssistiveTechnologies" />

## Disable text wrapping

To prevent text wrapping inside a button, apply the `.text-nowrap` class to the `<CButton>`. Alternatively, set `$btn-white-space: nowrap` in Sass to apply it globally.

## Button components

The `<CButton>` component can render as a `<button>`, `<a>`, or `<input>` element. When using it as an anchor for functionality (like toggles), add `role="button"` for accessibility.

<ExampleSnippet className="theme-bootstrap" component="ButtonComponentsExample" componentName="React Bootstrap Button" />

## Outline buttons

Use `variant="outline"` to create buttons with a transparent background and colored border — a more subtle alternative to standard buttons.

<ExampleSnippet className="theme-bootstrap" component="ButtonOutlineExample" componentName="React Bootstrap Button" />

## Ghost buttons

For minimal buttons with no background or border, use `variant="ghost"`. These are great for secondary actions.

<ExampleSnippet className="theme-bootstrap" component="ButtonGhostExample" componentName="React Bootstrap Button" />

<Callout color="info">
  Some button styles use light text colors and should be placed on dark backgrounds to maintain contrast.
</Callout>

## Sizes

Create large or small buttons with the `size="lg"` or `size="sm"` props.

<ExampleSnippet className="theme-bootstrap" component="ButtonSizesExample" componentName="React Bootstrap Button" />

<ExampleSnippet className="theme-bootstrap" component="ButtonSizes2Example" componentName="React Bootstrap Button" />

You can also use CSS variables for fully custom sizes.

<ExampleSnippet className="theme-bootstrap" component="ButtonSizes3Example" componentName="React Bootstrap Button" />

## Shapes

### Pill buttons

Use pill-shaped buttons for a soft, rounded appearance.

<ExampleSnippet className="theme-bootstrap" component="ButtonShapePillExample" componentName="React Bootstrap Button" />

### Square buttons

Use square-shaped buttons for compact, uniform interactions.

<ExampleSnippet className="theme-bootstrap" component="ButtonShapeSquareExample" componentName="React Bootstrap Button" />

## Disabled state

Disable any `<CButton>` by passing the `disabled` prop. This applies `pointer-events: none` and dims the button.

<ExampleSnippet className="theme-bootstrap" component="ButtonDisabledExample" componentName="React Bootstrap Button" />

Disabling anchor elements (`<a>`) requires additional steps: CoreUI applies the `.disabled` class, adds `aria-disabled="true"`, and sets `tabindex="-1"`.

<ExampleSnippet className="theme-bootstrap" component="ButtonDisabled2Example" componentName="React Bootstrap Button" />

These steps prevent users from interacting with disabled links via keyboard or mouse.

## Block buttons

To make buttons span the full width of their container, use Bootstrap's layout utilities like `.d-grid`, `.w-100`, and `.col-*`.

<ExampleSnippet className="theme-bootstrap" component="ButtonBlockExample" componentName="React Bootstrap Button" />

Responsive layout: vertically stacked until the `md` breakpoint, where `.d-md-block` reverts to inline.

<ExampleSnippet className="theme-bootstrap" component="ButtonBlock2Example" componentName="React Bootstrap Button" />

You can also create column-based buttons using `.col-6` and center them with `.mx-auto`.

<ExampleSnippet className="theme-bootstrap" component="ButtonBlock3Example" componentName="React Bootstrap Button" />

Use flex utilities and margin helpers to align buttons horizontally when they're no longer stacked.

<ExampleSnippet className="theme-bootstrap" component="ButtonBlock4Example" componentName="React Bootstrap Button" />

## API reference

See the full API reference for all available props in the Bootstrap-styled React Button component.

- [&lt;CButton /&gt;](./api/#cbutton)
