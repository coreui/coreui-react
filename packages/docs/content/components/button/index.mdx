---
title: React Button Component
name: Buttons
description: React button component for actions in tables, forms, cards, and more. CoreUI for React.js provides various styles, states, and size. Ready to use and easy to customize.
route: /components/button/
other_frameworks: button
---

## How to use React Button Component.

CoreUI includes a bunch of predefined buttons components, each serving its own semantic purpose. React buttons show what action will happen when the user clicks or touches it. CoreUI buttons are used to initialize operations, both in the background or foreground of an experience.

<ExampleSnippet component="ButtonExample" componentName="React Button" />

<Callout color="warning" type="colorAssistiveTechnologies" />

## Disable text wrapping

If you don't want the react button text to wrap, you can add the `.text-nowrap` className to the `<CButton>`. In Sass, you can set `$btn-white-space: nowrap` to disable text wrapping for each button.

## Button components

The `<CButton>` component are designed for `<button>` , `<a>` or `<input>` elements (though some browsers may apply a slightly different rendering).

If you're using `<CButton>` component as `<a>` elements that are used to trigger functionality ex. collapsing content, these links should be given a `role="button"` to adequately communicate their meaning to assistive technologies such as screen readers.

<ExampleSnippet component="ButtonComponentsExample" componentName="React Button" />

## Outline buttons

If you need a button, but without the strong background colors. Set `variant="outline"` prop to remove all background colors.

<ExampleSnippet component="ButtonOutlineExample" componentName="React Button" />

## Ghost buttons

If you need a ghost variant of react button, set `variant="ghost"` prop to remove all background colors.

<ExampleSnippet component="ButtonGhostExample" componentName="React Button" />

<Callout color="info">
  Some of the button styles use a relatively light foreground color, and should only be used on a
  dark background in order to have sufficient contrast.
</Callout>

## Sizes

Larger or smaller react buttons? Add `size="lg"` or `size="sm"` for additional sizes.

<ExampleSnippet component="ButtonSizesExample" componentName="React Button" />

<ExampleSnippet component="ButtonSizes2Example" componentName="React Button" />

You can even roll your own custom sizing with CSS variables:

<ExampleSnippet component="ButtonSizes3Example" componentName="React Button" />

## Shapes

### Pill buttons

<ExampleSnippet component="ButtonShapePillExample" componentName="React Button" />

### Square buttons

<ExampleSnippet component="ButtonShapeSquareExample" componentName="React Button" />

## Disabled state

Add the `disabled` boolean prop to any `<CButton>` component to make buttons look inactive. Disabled button has `pointer-events: none` applied to, disabling hover and active states from triggering.

<ExampleSnippet component="ButtonDisabledExample" componentName="React Button" />

Disabled buttons using the `<a>` component act a little different:

`<a>`s don't support the `disabled` attribute, so CoreUI has to add `.disabled` className to make buttons look inactive. CoreUI also has to add to the disabled button component `aria-disabled="true"` attribute to show the state of the component to assistive technologies.

<ExampleSnippet component="ButtonDisabled2Example" componentName="React Button" />

The `.disabled` class uses `pointer-events: none` to try to disable the link functionality of `<a>`s, but that CSS property is not yet standardized. Besides, even in browsers that do support `pointer-events: none`, keyboard navigation remains unaffected, meaning that sighted keyboard users and users of assistive technologies will still be able to activate these links. So to be safe, we automatically add a `tabindex="-1"` attribute on disabled links (to prevent them from receiving keyboard focus) and use custom JavaScript to disable their functionality.

## Block buttons

Create buttons that span the full width of a parentâ€”by using utilities.

<ExampleSnippet component="ButtonBlockExample" componentName="React Button" />

Here we create a responsive variation, starting with vertically stacked buttons until the `md` breakpoint, where `.d-md-block` replaces the `.d-grid` class, thus nullifying the `gap-2` utility. Resize your browser to see them change.

<ExampleSnippet component="ButtonBlock2Example" componentName="React Button" />

You can adjust the width of your block buttons with grid column width classes. For example, for a half-width "block button", use `.col-6`. Center it horizontally with `.mx-auto`, too.

<ExampleSnippet component="ButtonBlock3Example" componentName="React Button" />

Additional utilities can be used to adjust the alignment of buttons when horizontal. Here we've taken our previous responsive example and added some flex utilities and a margin utility on the button to right align the buttons when they're no longer stacked.

<ExampleSnippet component="ButtonBlock4Example" componentName="React Button" />

## API

Check out the documentation below for a comprehensive guide to all the props you can use with the components mentioned here.

- [&lt;CButton /&gt;](./api/#cbutton)
