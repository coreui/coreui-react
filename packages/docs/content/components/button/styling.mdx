---
title: React Button Component Styling
name: Button Styling
description: Learn how to customize the React Button component with CSS classes, variables, and SASS for flexible styling and seamless integration into your design.
route: /components/button/
---

### CSS variables

React Button supports CSS variables for easy customization. These variables are set via SASS but allow direct overrides in your stylesheets or inline styles.

<ScssDocs file="_buttons.scss" capture="btn-css-vars" />

#### How to use CSS variables

```jsx
const customVars = {
  '--cui-btn-color': '#ffe484',
  '--cui-btn-bg': '#efefef',
}

return <CButton style={customVars}>{/* Button content */}</CButton>
```

### SASS variables

<ScssDocs file="_variables.scss" capture="btn-variables" />
