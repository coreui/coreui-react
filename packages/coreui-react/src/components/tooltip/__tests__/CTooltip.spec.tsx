import * as React from 'react'
import { act, fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { CLink } from '../../link'
import { CTooltip } from '../index'
import { CButton } from '../../button'

let container: HTMLDivElement | null

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  container && container.remove()
  container = null
})

test('loads and displays CTooltip component', async () => {
  const { container } = render(
    <CTooltip content="content">
      <CLink>Test</CLink>
    </CTooltip>
  )
  expect(container).toMatchSnapshot()
})

test('CTooltip customize', async () => {
  jest.useFakeTimers()

  render(
    <CTooltip trigger="hover" placement="right" content="content">
      <CLink href="#">Test</CLink>
    </CTooltip>
  )

  const link = screen.getByText('Test')

  act(() => {
    fireEvent.mouseOver(link)
    jest.runAllTimers()
  })

  expect(document.querySelector('.tooltip')).toBeInTheDocument()

  jest.useRealTimers()
})

test('CTooltip onShow and onHide', async () => {
  jest.useFakeTimers()

  const onShow = jest.fn()
  const onHide = jest.fn()

  render(
    <CTooltip trigger="click" placement="right" content="content" onShow={onShow} onHide={onHide}>
      <CButton>Test</CButton>
    </CTooltip>
  )

  const btn = screen.getByRole('button', { name: /test/i })

  expect(onShow).toHaveBeenCalledTimes(0)
  expect(onHide).toHaveBeenCalledTimes(0)

  act(() => {
    fireEvent.click(btn)
    jest.runAllTimers()
  })

  expect(onShow).toHaveBeenCalledTimes(1)
  expect(onHide).toHaveBeenCalledTimes(0)

  act(() => {
    fireEvent.click(btn)
    jest.runAllTimers()
  })

  expect(onShow).toHaveBeenCalledTimes(1)
  expect(onHide).toHaveBeenCalledTimes(1)

  jest.useRealTimers()
})
