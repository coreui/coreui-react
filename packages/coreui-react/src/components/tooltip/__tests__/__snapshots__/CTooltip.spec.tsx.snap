// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTooltip customize 1`] = `
<div>
  <a
    class="link"
  >
    Test
  </a>
</div>
`;

exports[`loads and displays CTooltip component 1`] = `
<div>
  <a
    class=""
  >
    Test
  </a>
</div>
`;
